### DISCUSSION TITLE
How to correspond the  ```edge_attr``` to reorder ```new_edge_index```?

### DISCUSSION THREAD

**Discussion by jiaruHithub** [pyg-team/pytorch_geometric#5512] [Q&A] [answered]
I reorder a ```adj``` of a graph

```
adj = to_dense_adj(graph.edge_index)[0].numpy()
new_adj= adj [idx, :]
new_adj= new_adj[:, idx]
new_edge_index, _ = dense_to_sparse(A)
```
Now ```adj``` has new order, and this dataset has ```edge_attr```, the order of ```edge_attr``` is  correspond to ```edge_index```. My question is : 
How to correspond the  ```edge_attr``` to reorder ```new_edge_index```?
Thanks for aswner!!!

**Comment by EdisonLeeeee** [comment, marked as answer]
Did you intend to reorder both `edge_index` and `edge_attr`? If so, you can use `torch_geometric.utils.sort_edge_index` instead.

**Reply by jiaruHithub** [reply]
Thanks, how to use my own generated order ```own_index``` in ```torch_geometric.utils.sort_edge_index``` ? I didn't find the right example

**Reply by EdisonLeeeee** [reply]
There is no option to use your own `own_index` to sort your `edge_index`. Looking at your first example, it seems you are extracting a subgraph of `graph.edge_index` with a subset `idx`. If so, we can have a better solution using `torch_geometric.utils.subgraph` rather than `to_dense_adj`. Can you clarify?

**Reply by jiaruHithub** [reply]
thanks，it not a subgraph extraction. ```idx``` contains all node indexs, which is a new nodes order. This operation reorder the ```adj```. Because the order of ```adj``` is changed, I want to change the order of ```edge_attr``` for right aggregation

**Reply by EdisonLeeeee** [reply]
Oh I get your point now. You can do that:
```python
edge_order = torch.arange(graph.edge_index.size(1))
adj = to_dense_adj(graph.edge_index, edge_attr=edge_order)[0].numpy()
new_adj= adj [idx, :]
new_adj= new_adj[:, idx]
new_edge_index, new_edge_order = dense_to_sparse(new_adj)
```
Then you can reorder your `edge_attr` via `new_edge_order`.

**Reply by jiaruHithub** [reply]
Thanks, but I didn't got this point, ```adj = to_dense_adj(graph.edge_index, edge_attr=edge_order)[0].numpy()``` fed the original edge order as ```edge_attr```, what it do ? the ```adj``` contain other imformation ?

**Reply by jiaruHithub** [reply]
Additional, I found use this code, the size of ```edge_index``` will changed, always become ```[2, normal_size - 1]```, and the edge order is error, the start and end of the edge is shuffled. Like this toy:
```
pyg_graph = Data()
pyg_graph.x = torch.tensor([[-1], [0], [1]], dtype=torch.float)
pyg_graph.edge_index = torch.tensor([[0, 1, 1, 2],
                           [1, 0, 2, 1]], dtype=torch.long)
#pyg_graph.num_nodes = num_nodes
idx = [2,0,1]
edge_order = torch.arange(pyg_graph.edge_index.size(1))
adj = to_dense_adj(pyg_graph.edge_index, edge_attr=edge_order)[0].numpy()
print(adj)
new_adj= adj [idx, :]
new_adj= new_adj[:, idx]
print(new_adj)
new_edge_index, new_edge_order = dense_to_sparse(torch.from_numpy(new_adj))
print(new_edge_index, new_edge_order)
```
output is :
```
[[0 0 0]
 [1 0 2]
 [0 3 0]]
[[0 0 3]
 [0 0 0]
 [2 1 0]]
tensor([[0, 2, 2],
        [2, 0, 1]]) tensor([3, 2, 1])
```
so now the problem is the```adj``` seems not right, the edge number of ```new_edge_index``` is ```original_num-1```, and the ```new_edge_order``` is ```original_num-1```.

**Reply by EdisonLeeeee** [reply]
My bad. `edge_order` should be greater than zero before `dense_to_sparse` :
```python
edge_order = torch.arange(pyg_graph.edge_index.size(1)) + 1 # + 1 here

adj = to_dense_adj(pyg_graph.edge_index, edge_attr=edge_order)[0].numpy()
print(adj)
new_adj= adj [idx, :]
new_adj= new_adj[:, idx]
print(new_adj)
new_edge_index, new_edge_order = dense_to_sparse(torch.from_numpy(new_adj))

new_edge_order -= 1 # -1 to restore the edge orders
```

**Reply by jiaruHithub** [reply]
Thanks your help, but another question is that the values of ```adj``` is not 1, we can manually make it equal to 1.
Is it normal?

**Reply by EdisonLeeeee** [reply]
The values in `adj` indicate the edge orders `edge_order`. It is used to record your edge orders when node orders changed. So let's keep it just it as.

**Reply by jiaruHithub** [reply]
So thanks. Thanks to helpers like you in the community. Have fun
